src/ffi/w3c/credential.rs

```diff
@@ -0,0 +1,403 @@
+/// cred_def:   object handle pointing to the credential definition
+/// cred_p:     reference that will contain converted credential (in W3C form) instance pointer
+///
+/// # Returns
+/// Error code
+#[no_mangle]
+pub extern "C" fn anoncreds_credential_to_w3c(
+    cred: ObjectHandle,
+    cred_def: ObjectHandle,
+    cred_p: *mut ObjectHandle,
+) -> ErrorCode {
+    catch_error(|| {
+        check_useful_c_ptr!(cred_p);
+
+        let credential = cred.load()?;
+        let credential = credential.cast_ref::<Credential>()?;
+
+        let w3c_credential = credential_to_w3c(credential, cred_def.load()?.cast_ref()?)?;
+        let w3c_cred = ObjectHandle::create(w3c_credential)?;
+
+        unsafe { *cred_p = w3c_cred };
+        Ok(())
+    })
+}
+
+/// Convert credential in W3C form into legacy credential form
+///
+/// # Params
+/// cred:       object handle pointing to credential in W3C form to convert
+/// cred_p:     reference that will contain converted credential (in legacy form) instance pointer
+///
+/// # Returns
+/// Error code
+#[no_mangle]
+pub extern "C" fn anoncreds_credential_from_w3c(
+    cred: ObjectHandle,
+    cred_p: *mut ObjectHandle,
+) -> ErrorCode {
+    catch_error(|| {
+        check_useful_c_ptr!(cred_p);
+
+        let credential = cred.load()?;
+        let credential = credential.cast_ref::<W3CCredential>()?;
+
+        let credential = credential_from_w3c(credential)?;
+        let cred = ObjectHandle::create(credential)?;
+
+        unsafe { *cred_p = cred };
+        Ok(())
+    })
+}
+
+/// Add Non-Anoncreds Data Integrity proof signature to W3C AnonCreds credential
+///
+/// # Params
+/// cred:       object handle pointing to W3C AnonCreds credential
+/// proof:      data integrity proof as JSON string
+/// cred_p:     reference that will contain update credential
+///
+/// # Returns
+/// Error code
+#[no_mangle]
+pub extern "C" fn anoncreds_w3c_credential_add_non_anoncreds_integrity_proof(
+    cred: ObjectHandle,
+    proof: FfiStr,
+    cred_p: *mut ObjectHandle,
+) -> ErrorCode {
+    catch_error(|| {
+        check_useful_c_ptr!(cred_p);
+
+        let proof = proof
+            .as_opt_str()
+            .ok_or_else(|| err_msg!("Missing proof"))?;
+        let proof: NonAnonCredsDataIntegrityProof =
+            serde_json::from_str(proof).map_err(|_| err_msg!("Unable to parse proof"))?;
+
+        let mut cred = cred.load()?.cast_ref::<W3CCredential>()?.clone();
+
+        cred.add_non_anoncreds_integrity_proof(proof);
+
+        let cred = ObjectHandle::create(cred)?;
+        unsafe { *cred_p = cred };
+
+        Ok(())
+    })
+}
+
+/// Set id to W3C AnonCreds credential
+///
+/// # Params
+/// cred:       object handle pointing to W3C AnonCreds credential
+/// id:         id to add into credential
+/// cred_p:     reference that will contain update credential
+///
+/// # Returns
+/// Error code
+#[no_mangle]
+pub extern "C" fn anoncreds_w3c_credential_set_id(
```

Comment:
Is there a specific reason why we expose these set methods over FFI? IIRC we did not do that for the anoncreds credentials, is this a nice convience for w3c specifically after conversion or something?

Reply:
We added these helper methods to provide an ability to set Non-AnonCreds Data Integrity Proof Signatures for generated W3C credential objects. So that W3C credentials may contain multiple signatures to use in future.
It's up to application whether to use these methods or add proof by itself.